File: src/app/stdin.ts
import { TsqError } from "../errors";

const STDIN_TIMEOUT_MS = 30_000;

export async function readStdinContent(): Promise<string> {
  const readAll = async (): Promise<string> => new Response(Bun.stdin.stream()).text();

  const timeout = new Promise<never>((_resolve, reject) => {
    const timer = setTimeout(() => {
      reject(
        new TsqError(
          "VALIDATION_ERROR",
          `stdin read timed out after ${STDIN_TIMEOUT_MS / 1000} seconds`,
          1,
        ),
      );
    }, STDIN_TIMEOUT_MS);
    if (typeof timer === "object" && "unref" in timer) {
      timer.unref();
    }
  });

  const result = await Promise.race([readAll(), timeout]);
  if (result.trim().length === 0) {
    throw new TsqError("VALIDATION_ERROR", "stdin content must not be empty", 1);
  }
  return result;
}

File: src/app/stdin.rs
use crate::errors::TsqError;
use std::io::Read;
use std::sync::mpsc::{self, RecvTimeoutError};
use std::thread;
use std::time::Duration;

const STDIN_TIMEOUT_MS: u64 = 30_000;

pub fn read_stdin_content() -> Result<String, TsqError> {
    let (tx, rx) = mpsc::channel();

    thread::spawn(move || {
        let mut input = String::new();
        let result = std::io::stdin().read_to_string(&mut input).map(|_| input);
        let _ = tx.send(result);
    });

    match rx.recv_timeout(Duration::from_millis(STDIN_TIMEOUT_MS)) {
        Ok(result) => {
            let content = result.map_err(|error| {
                TsqError::new("IO_ERROR", "failed reading stdin", 2)
                    .with_details(serde_json::json!({"message": error.to_string()}))
            })?;
            if content.trim().is_empty() {
                return Err(TsqError::new(
                    "VALIDATION_ERROR",
                    "stdin content must not be empty",
                    1,
                ));
            }
            Ok(content)
        }
        Err(RecvTimeoutError::Timeout) => Err(TsqError::new(
            "VALIDATION_ERROR",
            "stdin read timed out after 30 seconds",
            1,
        )),
        Err(RecvTimeoutError::Disconnected) => {
            Err(TsqError::new("IO_ERROR", "failed reading stdin", 2))
        }
    }
}
